docs/docs/data-fetching.md

```diff
@@ -0,0 +1,221 @@
+
+```graphql
+query {
+  example {
+    nameWithOwner
+    url
+  }
+}
+```
+
+#### Writing a query to gather the static data needed for a page
+
+With the source plugin installed and added to your config, you can write a [static query](/docs/use-static-query/) that will retrieve the necessary data from Gatsby's data layer while building the site.
+
+```jsx:title=src/pages/index.js
+import React from "react"
+import { graphql, useStaticQuery } from "gatsby" // highlight-line
+
+const IndexPage = () => {
+  // highlight-start
+  const gatsbyRepoData = useStaticQuery(graphql`
+    query {
+      github {
+        repository(name: "gatsby", owner: "gatsbyjs") {
+          id
+          nameWithOwner
+          url
+        }
+      }
+    }
+  `)
+  // highlight-end
+
+  return (
+    <section>
+      <p>
+        Build Time Data: Gatsby repo{` `}
+        <a href={gatsbyRepoData.github.repository.url}>
+          {gatsbyRepoData.github.repository.nameWithOwner} // highlight-line
+        </a>
+      </p>
+    </section>
+  )
+}
+
+export default IndexPage
+```
+
+> This data is gathered at build time and written to a JSON file, the code is rewritten as the site is built to import the JSON file and set `gatsbyRepoData` equal to the contents of the JSON file instead of the call to `useStaticQuery`, you can read more about this process in the Gatsby internals section on [Normal vs Static Queries](/docs/static-vs-normal-queries/#replacing-queries-with-json-imports)
+
+Here's an adaptation of this build time data example being used on this page:
+
+<BuildDataExample />
+
+> the linked url and repository name are fetched at build time, if the name of the repository changed and the site were rebuilt it would change
+
+### Fetching data at client-side runtime
+
+For fetching data at runtime, you can use any method to retrieve data that you would use in a regular React app.
+
+#### Retrieving data with the `fetch` API
+
+The `fetch` API is a modern implementation of the older, well-supported `XMLHttpRequest` (also known as AJAX).
+
+With the `useState` and `useEffect` hooks from React, you can query for the data once when the page loads, and save the data returned to a variable called `starsCount`. Every time the page is refreshed, `fetch` will go to the GitHub API to retrieve the most up-to-date version of the data.
+
+```jsx:title=src/pages/index.js
+import React, { useState, useEffect } from "react" // highlight-line
+import { graphql, useStaticQuery } from "gatsby"
+
+const IndexPage = () => {
+  // Build Time Data Fetching
+  const gatsbyRepoData = useStaticQuery(graphql`
+    query {
+      github {
+        repository(name: "gatsby", owner: "gatsbyjs") {
+          id
+          nameWithOwner
+          url
+        }
+      }
+    }
+  `)
+  // Client-side Runtime Data Fetching
+  // highlight-start
+  const [starsCount, setStarsCount] = useState(0)
+  useEffect(() => {
+    const fetchData = async () => {
```

Comment:
I'm unsure about this syntax, as I don't think many developers new to react will understand why you need to create a sub-function for this. It's not explained in the reactjs.org docs either so I feel like we should add a comment or convert it to a promise chain like so:

```js
fetch(`https://api.github.com/repos/gatsbyjs/gatsby`)
  .then(response => respone.json())
  .then(resultData => { setStarsCount(resultData.stargazers_count) })
```

Reply:
I don't see any problem with that, true that it would be easier to grasp than having to have the React understanding beforehand. I'll update it 🙂